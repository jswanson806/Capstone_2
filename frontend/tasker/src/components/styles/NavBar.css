.Navbar .container-fluid {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: #c5dacc;
    font-family: 'Raleway', sans-serif;
    font-weight: bold;
  }
  
  .Navbar .Navbar-nav {
    display: flex;
    list-style: none;
    margin: 0;
    padding: 0;
  }
  
  .Navbar .Navbar-nav li {
    margin-right: 20px;
  }
  
  .Navbar .navbar-nav li:last-child {
    margin-right: 0;
  }
  
  .Navbar .Navlink.active {
    color: #f1bb7b;
  }

  /* .Navbar .Navlink-button {
    background: none;
    border: none;
    font-family: 'Raleway', sans-serif;
    font-weight: bold;
    font-size: small;
    color: #0f5959;
    text-decoration: none;
    cursor: pointer;
  } */