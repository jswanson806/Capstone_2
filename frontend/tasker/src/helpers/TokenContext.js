import React, { createContext, useState } from 'react';

// Create the context
const TokenContext = createContext();

// Create a provider for the context
const TokenProvider = ({ children }) => {
  const [token, setToken] = useState(localStorage.getItem('token') || '');

  // Function to update the token in the state and localStorage
  const updateToken = (newToken) => {
    setToken(newToken);
    localStorage.setItem('token', newToken);
  };

  return (
    <TokenContext.Provider value={{ token, updateToken }}>
      {children}
    </TokenContext.Provider>
  );
};

export { TokenContext, TokenProvider };
