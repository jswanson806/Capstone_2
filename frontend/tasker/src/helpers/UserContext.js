import React, { createContext, useState } from 'react';

// Create the context
const UserContext = createContext();

// Create a provider for the context
const UserProvider = ({ children }) => {
  const [user, setUser] = useState(localStorage.getItem('user') || '');

  // Function to update the user in the state and localStorage
  const updateUser = (newUser) => {
    setUser(newUser);
    localStorage.setItem('user', newUser);
  };

  return (
    <UserContext.Provider value={{ user, updateUser }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
