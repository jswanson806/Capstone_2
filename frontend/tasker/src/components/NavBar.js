import React, { useContext, useState } from "react";
import { NavLink, useNavigate } from "react-router-dom";
import { Navbar, Nav, NavItem, NavbarBrand, NavbarToggler, Collapse, Button } from 'reactstrap';
import { TokenContext } from "../helpers/TokenContext";
import { UserContext } from "../helpers/UserContext";
import "./styles/NavBar.css";


const NavBar = () => {

    const navigate = useNavigate();

    const { token, updateToken } = useContext(TokenContext);
    const { updateUser } = useContext(UserContext);
    const [collapsed, setCollapsed] = useState(true);

    let loggedIn = (token && token !== '') ? true : false;

    const toggleNavbar = () => setCollapsed(!collapsed);

    /** Removes token and user from localStorage 
     * 
     * Redirects to /login
    */
    const logOut = () => {
        updateToken('');
        updateUser('');
        navigate("/login");
    }

    return(
        <div>
        <Navbar container="xl" expand="md" color="dark" dark data-testid="Navbar" >
            {!loggedIn ? (
                <NavLink to="/" className="me-auto" >
                    <Button color="info" outline data-testid="Navbar-tasker">Tasker</Button>
                </NavLink>
            ) : (
                <NavLink to="/jobs" >
                    <Button color="info" outline data-testid="Navbar-dashboard">Tasker</Button>
                </NavLink>
            )}

            <NavbarToggler onClick={toggleNavbar} className="me-2"/>
            <Collapse isOpen={!collapsed} navbar>
                <Nav navbar className="ml-auto">

                    {/* Conditionally render user navigation links based on loggedIn status */}
                    {!loggedIn ? (
                        <>
                            <NavItem className="mr-2">
                                <NavLink to="/login" >
                                    <Button color="warning" outline data-testid="Navbar-login">Login</Button>
                                </NavLink>
                            </NavItem>
                            <NavItem className="mr-2">
                                <NavLink to="/signup" >
                                    <Button color="info" outline data-testid="Navbar-signup">Sign Up</Button>
                                </NavLink>
                            </NavItem>
                        </>
                    ) : (
                        <>
                            <NavItem className="mr-2">
                                <NavLink to="/jobs" >
                                    <Button color="info" outline data-testid="Navbar-jobs">Jobs</Button>
                                </NavLink>
                            </NavItem>
                            <NavItem className="mr-2">
                                <NavLink to="/messages">
                                    <Button color="info" outline data-testid="Navbar-messages">Messages</Button>
                                </NavLink>
                            </NavItem>
                            <NavItem className="mr-2">
                                <Button onClick={logOut} data-testid="Navbar-logout-button" color="danger" outline>Logout</Button>
                            </NavItem>
                        </>
                    )}

                </Nav>
            </Collapse>

        </Navbar>
        </div>
    )
};

export default NavBar;