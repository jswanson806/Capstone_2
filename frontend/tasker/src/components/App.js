import React from 'react';
import Router from './Router.js';
import NavBar from './NavBar.js';
import './styles/App.css';
import { TokenProvider } from '../helpers/TokenContext.js';

/** App Component
 * 
 * Handles the rendering of the NavBar and Router components.
 * 
 * Uses TokenProvider to provide global access to the token
 * 
 */

function App() {

  return (
        <TokenProvider>
          <NavBar />
          <Router />
        </TokenProvider>
  );
}

export default App;