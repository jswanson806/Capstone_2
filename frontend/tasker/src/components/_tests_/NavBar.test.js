import React from "react";
import { render, act, waitFor, screen, fireEvent, queryByTestId} from '@testing-library/react';
import NavBar from "../NavBar.js";
import App from "../App.js";
import "reactstrap";
import { TokenContext } from "../../helpers/TokenContext.js";
import TaskerApi from "../../api.js";
import { MemoryRouter } from "react-router-dom";
import { UserContext } from "../../helpers/UserContext.js";


describe('smoke and snapshot tests', () => {

    test('renders without crashing', async () => {
        render(
            <MemoryRouter >
                <UserContext.Provider value={UserContext} >
                <TokenContext.Provider value={TokenContext}>
                    <NavBar />
                </TokenContext.Provider >
                </UserContext.Provider>
            </MemoryRouter>
        )
    });


    test('matches snapshot', async () => {
        const {asFragment} = render(
            <MemoryRouter >
                <UserContext.Provider value={UserContext} >
                <TokenContext.Provider value={TokenContext}>
                    <NavBar />
                </TokenContext.Provider >
                </UserContext.Provider>
            </MemoryRouter>
        )

        expect(asFragment()).toMatchSnapshot();
    });
});


describe('renders correct content', () => {
    test('works - not logged in', async () => {

        expect.assertions(6);

        render(
            <MemoryRouter >
                <UserContext.Provider value={UserContext} >
                <TokenContext.Provider value={''}> 
                    <NavBar />
                </TokenContext.Provider >
                </UserContext.Provider>
            </MemoryRouter>
        );

        const taskerLink = screen.queryByTestId("Navbar-tasker");
        const logoutBtn = screen.queryByTestId("Navbar-logout-button");
        const jobsLink = screen.queryByTestId("Navbar-jobs");
        const messagesLink = screen.queryByTestId("Navbar-messages");
        const loginLink = screen.queryByTestId("Navbar-login");
        const signupLink = screen.queryByTestId("Navbar-signup");

        // token context set to empty string results in 'Tasker' button render
        expect(taskerLink).toBeInTheDocument();
        expect(loginLink).toBeInTheDocument();
        expect(signupLink).toBeInTheDocument();
        expect(jobsLink).not.toBeInTheDocument();
        expect(messagesLink).not.toBeInTheDocument();

        // should render logOut button
        expect(logoutBtn).not.toBeInTheDocument();
    });

    test('works - logged in', async () => {

        expect.assertions(6);

        render(
            <MemoryRouter >
                <UserContext.Provider value={UserContext} >
                <TokenContext.Provider value={{ token: 'test-token', updateToken: () => {} }}> 
                    <NavBar />
                </TokenContext.Provider >
                </UserContext.Provider>
            </MemoryRouter>
        );

        const taskerLink = screen.queryByTestId("Navbar-tasker");
        const logoutBtn = screen.queryByTestId("Navbar-logout-button");
        const jobsLink = screen.queryByTestId("Navbar-jobs");
        const messagesLink = screen.queryByTestId("Navbar-messages");
        const loginLink = screen.queryByTestId("Navbar-login");
        const signupLink = screen.queryByTestId("Navbar-signup");

        // token context not being an empty string results in 'Dashboard' link render
        expect(taskerLink).not.toBeInTheDocument();
        expect(loginLink).not.toBeInTheDocument();
        expect(signupLink).not.toBeInTheDocument();
        expect(jobsLink).toBeInTheDocument();
        expect(messagesLink).toBeInTheDocument();
        expect(logoutBtn).toBeInTheDocument();
        
    });
});