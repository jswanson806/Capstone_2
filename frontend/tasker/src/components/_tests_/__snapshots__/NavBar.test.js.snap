// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="navbar navbar-expand-md navbar-dark bg-dark"
      container="xl"
      data-testid="Navbar"
    >
      <a
        aria-current="page"
        class="me-auto active"
        href="/"
      >
        <button
          class="btn btn-outline-info"
          data-testid="Navbar-tasker"
        >
          Tasker
        </button>
      </a>
      <button
        aria-label="Toggle navigation"
        class="me-2 navbar-toggler"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="collapse navbar-collapse"
      >
        <ul
          class="ml-auto navbar-nav"
        >
          <li
            class="mr-2 nav-item"
          >
            <a
              class=""
              href="/login"
            >
              <button
                class="btn btn-outline-warning"
                data-testid="Navbar-login"
              >
                Login
              </button>
            </a>
          </li>
          <li
            class="mr-2 nav-item"
          >
            <a
              class=""
              href="/signup"
            >
              <button
                class="btn btn-outline-info"
                data-testid="Navbar-signup"
              >
                Sign Up
              </button>
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </div>
</DocumentFragment>
`;
